### DISCUSSION TITLE
UFS .settings file corrupted

### DISCUSSION THREAD

**Discussion by UBWH** [arendst/Tasmota#20671] [Q&A] [answered]
Seems my .settings file is corrupted and I don't know how to fix it.

[EDIT] ESP32 with Tasmota 13.2.0

## BACKGROUND ##
Using SCRIPTing, with permanent variables.
Noticed that after ever reboot, if I display some permanent variables on a webpage, that the string contained a snipet from the script code (!) instead of the correct string.
![image](https://github.com/arendst/Tasmota/assets/72185209/6c966e4e-9bea-4625-bbb5-722b35a59483)

I had a look at the (binary) UFS `.settings` file and can see this:
![Screenshot 2024-02-06 085010](https://github.com/arendst/Tasmota/assets/72185209/26fbd85a-9a10-46b4-8cc4-5cf013c6777e)

I have tried saving the corrupt `p:` variable, but the problem re-occurs on every reboot.
```
>D 48
p:aet="" ;permanent To    <---- this is corrupted
aett=""   ;temp To
...

;Every time MAIN page is displayed
aett=aet  ;Copy permanent to temp

;This executes when user presses  [SAVE] button on sub-webpage
if chg[ase]>0   
and ase==1 {
 aet=aett
 svars
}

; Extract from the sub-webpage
<tr><th>To<td>
tx(aett "")

```

## WHEN DID THIS OCCUR? ##
Can't be sure but I think (?) it may have been when I SAVED my script after editing - and Tasmota crashed and came back with a RESTART REASON of "Excess CPU Load" (or similar)

## QUESTION ##
How do I reset the .settings file to remove the corruption?

**Comment by joba-1** [comment]
Wepbage: Tools / Manage File system, Show hidden files, Select 🔥 next to .settings

But it could be some autoexec file or boot rule that recreates it.

P.S. not tested, could set your ESP on fire… ;)

**Reply by UBWH** [reply]
Sure - I can simply delete the .settings file .... but that sounds very dangerous. Not sure what bad thing will happen.

Can anyone confirm OK to simply delete `.settings`?

**Comment by gemu2015** [comment, marked as answer]
actually permanent string variables are not recommended at all. i will change the docs accordingly.
if you have a file system available it is the best strategy to have your own settings file. you can inspect, change or delete the file as you want.

simple example:

```
>D
fres=0
str="hello world"
numvar1=1.2345
tstr=""

>B

=#save_settings
=#load_settings
print str = %str%
print numvar1 = %4numvar1%

#save_settings
fres=fo("/mysettings.txt" w)
if fres>=0 {
	fw(str+"\n" fres)
	fw(s(4numvar1)+"\n" fres)
	fc(fres)
}

#load_settings
fres=fo("/mysettings.txt" r)
if fres>=0 {
	fr(str fres)
	fr(tstr fres)
	numvar1=tstr
	fc(fres)
}
```

**Reply by UBWH** [reply]
@gemu 
OK - I can make my own mysettings file.
I notice in the example, you are saving a string, and a float (as a string)

Do you also not recommend permanent `p:` vars for

- integers?
- floats?

**Reply by gemu2015** [reply]
floats and integers are no problem. in most cases strings are also OK.
as all permanent vars are stored sequentially you should always add new p vars at the end, otherwise  they are mixed up an in case of strings destroyed.

**Comment by UBWH** [comment]
`you should always add new p vars at the end`
Bingo!

That could explain everything. I DID add a new p:string_var in the middle of my list of `p:` vars

Was not aware of this restriction.
:-)

**Reply by gemu2015** [reply]
i just now added this to the documentation
